core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnDelta.scala

```diff
@@ -165,4 +165,50 @@ object ColumnDelta {
    */
   def deltaColumnIndex(tableColumnIndex: Int, hierarchyDepth: Int): Int =
     -tableColumnIndex * MAX_DEPTH + ColumnFormatEntry.DELETE_MASK_COL_INDEX - 1 - hierarchyDepth
+
+  /**
+   * Check if all the rows in a batch have been deleted.
+   */
+  private[columnar] def checkBatchDeleted(deleteBuffer: ByteBuffer): Boolean = {
+    val allocator = ColumnEncoding.getAllocator(deleteBuffer)
+    val bufferBytes = allocator.baseObject(deleteBuffer)
+    val bufferCursor = allocator.baseOffset(deleteBuffer) + 4
+    val numBaseRows = ColumnEncoding.readInt(bufferBytes, bufferCursor)
+    val numDeletes = ColumnEncoding.readInt(bufferBytes, bufferCursor + 4)
+    numDeletes >= numBaseRows
+  }
+
+  /**
+   * Delete entire batch from column store for the batchId and partitionId
+   * matching those of given key.
+   */
+  private[columnar] def deleteBatch(key: ColumnFormatKey, columnTableName: String,
+      columnRegion: Region[_, _], forUpdate: Boolean): Unit = {
+
+    // delete all the rows with matching batchId
+    def destroyKey(key: ColumnFormatKey): Unit = {
+      try {
+        columnRegion.destroy(key)
+      } catch {
+        case _: EntryNotFoundException => // ignore
+      }
+    }
+
+    val numColumns = key.getNumColumnsInTable(columnTableName)
+    // delete the stats rows first
+    destroyKey(key.withColumnIndex(ColumnFormatEntry.STATROW_COL_INDEX))
+    destroyKey(key.withColumnIndex(ColumnFormatEntry.DELTA_STATROW_COL_INDEX))
+    // column values and deltas next
+    for (columnIndex <- 1 to numColumns) {
```

Comment:
What will happen if the cache is getting closed when delete is inside this loop ? Some entries will be left behind ?

Reply:
This entire set of operations is under the same transaction as original operation, so if operation fails due to node failure then ideally entire thing should be rolled back. We are not there yet for persistent transactions of course. Otherwise do you have any solution?

Reply:
That's fine. I somehow got confused it won't be under transaction.